src/core/qgsmaplayer.cpp

```diff
@@ -1980,3 +1989,26 @@ void QgsMapLayer::onNotifiedTriggerRepaint( const QString &message )
   if ( refreshOnNotifyMessage().isEmpty() || refreshOnNotifyMessage() == message )
     triggerRepaint();
 }
+
+QgsRectangle QgsMapLayer::geographicExtent( bool actual ) const
```

Comment:
Shouldn't be more efficient to store the geographic extent instead of recomputing it every time this method is called?
I understand that it requires more work to keep it in sync with `mExtent` but I think it's a better approach.

Reply:
> I understand that it requires more work to keep it in sync with mExtent but I think it's a better approach.

I agree but the `mExtent` attribute is directly updated by subclasses in some cases, meaning that the `QgsMapLayer::setExtent()` method is not always called. So I don't see an easy way to keep the geographic extent in sync with the `mExtent` without inconsistencies.

>  Shouldn't be more efficient to store the geographic extent instead of recomputing it every time this method is called?

We could add a dedicated attribute `mOriginalXmlGeographicExtent` to store the original geographic extent read from the project and keep the `mGeographicExtent` up-to-date (in a lazy mode). It would be consistent with the `mOriginalXmlProperties` mechanism and it will avoid to waste cpu cycles.

Reply:
You could make `mExtent` private -- that's fine to do since it doesn't break python api, and would nicely force all the subclasses to use the correct setter

Reply:
> You could make `mExtent` private

Hi Nyall. 

Actually it's the first thing I tried but `mExtent` is also updated in the `QgsVectorLayer::extent()` const method. So I can't call `setExtent()` (not const) to update the `mExtent` private attribute.

And I didn't dig further because I didn't see how to rework `QgsVectorLayer::extent()` without risk. Unless I missed something?

Reply:
You could possibly get around that by making QgsVectorLayer a friend class. Or alternatively have a non const "setVectorLayerInternal" protected method which is hidden from python?

Reply:
@nyalldawson @elpaso what do you think of the last commit https://github.com/qgis/QGIS/pull/40444/commits/6dd07f01064a9c0ee3dcb405dc6002e6d68f2aea?